persist/fs/commitlog/commit_log.go

```diff
@@ -0,0 +1,275 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package commitlog
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3x/metrics"
+	"github.com/m3db/m3x/time"
+)
+
+const (
+	commitLogWriteQueue = 32768 // at 1MM write/s allow to fall behind by ~6.5% our deadline
+)
+
+var (
+	errCommitLogClosed = errors.New("commit log is closed")
+
+	writeRetries    = 2
+	writeRetryDelay = time.Millisecond
+	timeNow         = time.Now
+	timeZero        = time.Time{}
+)
+
+type completionFn func(err error)
+
+type commitLog struct {
+	sync.RWMutex
+	opts    m3db.DatabaseOptions
+	metrics xmetrics.Scope
+	nowFn   m3db.NowFn
+	// TODO(r): replace buffered channel with concurrent striped
+	// circular buffer to avoid central write lock contention
+	writes         chan commitLogWrite
+	writer         commitLogWriter
+	pendingAcks    []completionFn
+	bitset         bitset
+	writerExpireAt time.Time
+	closed         bool
+	closeErr       chan error
+}
+
+type commitLogWrite struct {
+	series       m3db.CommitLogSeries
+	datapoint    m3db.Datapoint
+	unit         xtime.Unit
+	annotation   []byte
+	attempt      int
+	completionFn completionFn
+}
+
+// NewCommitLog creates a new commit log
+func NewCommitLog(opts m3db.DatabaseOptions) (m3db.CommitLog, error) {
+	l := &commitLog{
+		opts:     opts,
+		nowFn:    opts.GetNowFn(),
+		writes:   make(chan commitLogWrite, commitLogWriteQueue),
+		metrics:  opts.GetMetricsScope().SubScope("commitlog"),
+		closeErr: make(chan error),
+	}
+
+	// Open the buffered commit log writer
+	if err := l.openWriter(l.nowFn()); err != nil {
+		return nil, err
+	}
+
+	// Flush the info header to ensure we can write to disk
+	if err := l.writer.Flush(); err != nil {
+		return nil, err
+	}
+
+	// Asynchronously write
+	go l.write()
+
+	return l, nil
+}
+
+func (l *commitLog) write() {
+	// TODO(r): also periodically flush the underlying commit log writer
+	// in case writes stall and commit log not written
+	var (
+		write commitLogWrite
+		retry = commitLogWrite{attempt: writeRetries}
+		open  bool
+	)
+	for {
+		if retry.attempt < writeRetries {
```

Comment:
Seems like if we have a write error, the write will be retried, but if the retried write succeeds, we'll retry the write here until `retry.attempt == writeRetries`? Might need to either do `retry.attempt = writeRetries`, or make the retry more explicit by having something along the lines of a `shouldRetry` flag and setting that to true when an error occurs. Right now we are implicitly using `attempt` to indicate whether we should retry, which I don't quite like.